src/utils.jl

```diff
@@ -94,3 +123,171 @@ function is_doc_expr(ex)
+        codeloc = code.codelocs[idx]
+    end
+    return codeloc
+end
+
+"""
+    stmtidx = statementnumber(frame, line)
+
+Return the index of the first statement in `frame`'s `CodeInfo` that corresponds to
+static line number `line`.
+"""
+function statementnumber(framecode::FrameCode, line)
+    lineidx = searchsortedfirst(framecode.src.linetable, line; by=lin->isa(lin,Integer) ? lin : lin.line)
+    1 <= lineidx <= length(framecode.src.linetable) || throw(ArgumentError("line $line not found in $(framecode.scope)"))
+    return searchsortedfirst(framecode.src.codelocs, lineidx)
+end
+statementnumber(frame::Frame, line) = statementnumber(frame.framecode, line)
+
+"""
+    framecode, stmtidx = statementnumber(method, line)
+
+Return the index of the first statement in `framecode` that corresponds to the given
+static line number `line` in `method`.
+"""
+function statementnumber(method::Method, line; line1=whereis(method)[2])
+    linec = line - line1 + method.line  # line number at time of compilation
+    framecode = get_framecode(method)
+    return framecode, statementnumber(framecode, linec)
+end
+
+## Printing
+
+function framecode_lines(src::CodeInfo)
+    buf = IOBuffer()
+    show(buf, src)
+    code = filter!(split(String(take!(buf)), '\n')) do line
+        !(line == "CodeInfo(" || line == ")" || isempty(line))
+    end
+    code .= replace.(code, Ref(r"\$\(QuoteNode\((.+?)\)\)" => s"\1"))
+    return code
+end
+framecode_lines(framecode::FrameCode) = framecode_lines(framecode.src)
+
+breakpointchar(framecode, stmtidx) =
+    isassigned(framecode.breakpoints, stmtidx) ? breakpointchar(framecode.breakpoints[stmtidx]) : ' '
+
+function print_framecode(io::IO, framecode::FrameCode; pc=0, range=1:nstatements(framecode), kwargs...)
+    iscolor = get(io, :color, false)
+    ndstmt = ndigits(nstatements(framecode))
+    lt = framecode.src.linetable
+    offset = lineoffset(framecode)
+    ndline = isempty(lt) ? 0 : ndigits(lt[end].line + offset)
+    nullline = " "^ndline
+    code = framecode_lines(framecode)
+    isfirst = true
+    for (stmtidx, stmtline) in enumerate(code)
+        stmtidx ∈ range || continue
+        bpc = breakpointchar(framecode, stmtidx)
+        isfirst || print(io, '\n')
+        isfirst = false
+        print(io, bpc, ' ')
+        if iscolor
+            color = stmtidx == pc ? Base.warn_color() : :normal
+            printstyled(io, lpad(stmtidx, ndstmt); color=color, kwargs...)
+        else
+            print(io, lpad(stmtidx, ndstmt), stmtidx == pc ? '*' : ' ')
+        end
+        line = linenumber(framecode, stmtidx)
+        print(io, ' ', line === nothing ? nullline : lpad(line, ndline), "  ", stmtline)
+    end
+end
+
+"""
+    local_variables = locals(frame::Frame)::Vector{Variable}
+
+Return the local variables as a vector of `Variable`[@ref].
+"""
+function locals(frame::Frame)
+    vars = Variable[]
+    data, code = frame.framedata, frame.framecode
+    for (sym, idx) in data.last_reference
```

Comment:
Not related to the PR but one thing I thought about here is that since `data.last_reference` is a `Dict` the order is unspecified. So perhaps we should iterate through the slot symbols, use a Set to keep track of what symbol we have added, but look up the value of that symbol in `data.last_reference`.

Reply:
I'll fix it here.

I'm also adding an `O(N^2)` performance note to `handle_err`; I propose we not worry about it until we detect a problem in practice.